include/seqan3/core/copyable_wrapper.hpp

```diff
@@ -0,0 +1,248 @@
+     * `t` needs to be callable with the passed arguments.
+     *
+     * Constness of this function depends on the constness of the call-operator of the wrapped object.
+     */
+    template <typename... args_t>
+        requires std::invocable<t, args_t...>
+    constexpr decltype(auto) operator()(args_t... args) noexcept(std::is_nothrow_invocable_v<t, args_t...>)
+    {
+        return std::invoke(this->value(), std::forward<args_t>(args)...);
+    }
+
+    //!\overload
+    template <typename... args_t>
+        requires std::invocable<t, args_t...>
+    constexpr decltype(auto) operator()(args_t... args) const noexcept(std::is_nothrow_invocable_v<t, args_t...>)
+    {
+        return std::invoke(this->value(), std::forward<args_t>(args)...);
+    }
+};
+
+/*!\brief Utility wrapper that behaves like std::optional but makes the type conform with the std::copyable concept.
+ * \ingroup core
+ * \see https://en.cppreference.com/w/cpp/ranges/copyable_wrapper
+ * \details
+ *
+ * If `t` is `std::copyable`, the STL allows to only store `t`, i.e. no `std::optional` is needed.
+ */
+template <boxable t>
+    requires std::copyable<t> || (std::is_nothrow_move_constructible_v<t> && std::is_nothrow_copy_constructible_v<t>)
+#if SEQAN3_WORKAROUND_DEFAULT_CONSTRUCTIBLE_VIEW // If views must be default constructible, t must also be
+    && std::default_initializable<t>
+#endif
+class copyable_wrapper<t>
+{
+private:
+    t value{}; //!< An object of the wrapped type.
+
+public:
+    //!\brief copyable_wrapper is default constructible, iff the wrapped type is default initialisable.
+    constexpr copyable_wrapper() requires std::default_initializable<t> = default;
+
+    constexpr copyable_wrapper(copyable_wrapper const &) = default; //!< Defaulted.
+    constexpr copyable_wrapper(copyable_wrapper &&) = default; //!< Defaulted.
+    constexpr ~copyable_wrapper() = default; //!< Defaulted.
+
+    //!\brief Copy assignment for copyable types is the default copy assignment.
+    constexpr copyable_wrapper & operator=(copyable_wrapper const &) requires std::copyable<t> = default;
+
+    //!\brief Copy assignment for non-copyable types uses the Destroy-then-copy paradigm.
+    constexpr copyable_wrapper & operator=(copyable_wrapper const & other) noexcept
+    {
+        // Destroy-then-copy
+        if (this != std::addressof(other))
+        {
+            value.~t(); // Call destructor of value
+            std::construct_at(std::addressof(value), *other); // Copy construct
+        }
+        return *this;
+    }
+
+    //!\brief Move assignment for copyable types is the default move assignment.
+    constexpr copyable_wrapper & operator=(copyable_wrapper &&) requires std::copyable<t> = default;
+
+    //!\brief Move assignment for non-copyable types uses the Destroy-then-copy paradigm.
+    constexpr copyable_wrapper & operator=(copyable_wrapper && other) noexcept
```

Comment:
Shouldn't the `noexcept` statement also depend on `std::is_nothrow_move_constructible_v<T>`?

Reply:
See above